aptos-move/move-examples/drand/sources/drand.move

```diff
@@ -0,0 +1,136 @@
+///     "period": 3,
+///     "genesis_time": 1677685200,
+///     "hash": "dbd506d6ef76e5f386f41c651dcb808c5bcbd75471cc4eafa3f4df7ad4e4c493",
+///     "groupHash": "a81e9d63f614ccdb144b8ff79fbd4d5a2d22055c0bfe4ee9a8092003dab1c6c0",
+///     "schemeID": "bls-unchained-on-g1",
+///     "metadata": {"beaconID": "fastnet"}
+/// }
+/// ```
+
+module drand::drand {
+    use std::hash::{sha3_256, sha2_256};
+    use std::option::{Self, Option, extract};
+    use std::vector;
+    use std::error;
+    use aptos_std::crypto_algebra::{eq, pairing, one, deserialize, hash_to, from_u64, serialize};
+    use aptos_std::bls12381_algebra::{G1, G2, Gt, FormatG2Compr, FormatG1Compr, HashG1XmdSha256SswuRo, Fr, FormatFrMsb};
+
+    /// The `bls-unchained-on-g1` drand beacon produces an output every 3 seconds.
+    /// (Or goes into catchup mode, if nodes fall behind.)
+    const PERIOD_SECS : u64 = 3;
+
+    /// The UNIX time (in seconds) at which the beacon started operating (this is the time of round #1)
+    const GENESIS_TIMESTAMP : u64 = 1677685200;
+
+    /// The drand beacon's PK, against which any beacon output for a round `i` can be verified.
+    const DRAND_PUBKEY : vector<u8> = x"a0b862a7527fee3a731bcb59280ab6abd62d5c0b6ea03dc4ddf6612fdfc9d01f01c31542541771903475eb1ec6615f8d0df0b8b6dce385811d6dcf8cbefb8759e5e616a3dfd054c928940766d9a5b9db91e3b697e5d70a975181e007f87fca5e";
+
+    /// The domain-separation tag (DST) used in drand's BLS signatures
+    const DRAND_DST: vector<u8> = b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_";
+
+    /// Error code for when anyone submits an incorrect randomness in our APIs (e.g., wrong-size).
+    const E_INCORRECT_RANDOMNESS: u64 = 1;
+
+    /// A randomness object, created from a successfully-verified piece of drand randomness.
+    /// This object can be converted into a uniform random integer.
+    struct Randomness has drop {
+        bytes: vector<u8>
+    }
+
+    /// Checks if the randomness in `bytes` verifies for the specified `round`.
+    /// If it verifies, returns the actual randomness, which is a hash function applied over `bytes`.
+    public fun verify_and_extract_randomness(
+        bytes: vector<u8>,
+        round: u64): Option<Randomness>
+    {
+        let pk = extract(&mut deserialize<G2, FormatG2Compr>(&DRAND_PUBKEY));
+        let sig = extract(&mut deserialize<G1, FormatG1Compr>(&bytes));
+        let msg_hash = hash_to<G1, HashG1XmdSha256SswuRo>(&DRAND_DST, &round_number_to_bytes(round));
+        assert!(eq(&pairing<G1, G2, Gt>(&msg_hash, &pk), &pairing<G1, G2, Gt>(&sig, &one<G2>())), 1);
+        option::some(Randomness {
+            bytes: sha3_256(bytes)
+        })
+    }
+
+    /// Returns a number in $[0, max)$ given some drand (verified) `randomness` and a `Type`.
+    /// Note: This is a one-shot API that consumes the `randomness`.
+    public fun random_number(randomness: Randomness, max: u64): u64 {
```

Comment:
probably you want to add a sequence number, now you only get one random number. I think there are plenty of of cases where you'd want to generate more random numbers

Reply:
Or perhaps &mut Randomness and update the randomness according to some reasonable pseudo-random algo

Reply:
Yes. But this is just a specific example for a lottery where only one random # is needed. Trying to get this out there for now. More complicated usecases require a fancier `Randomness` interface. See [AIP-41](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-41.md) for such a proposal.